contracts/exchange/contracts/src/MixinMatchOrders.sol

```diff
@@ -52,85 +117,182 @@ contract MixinMatchOrders is
 
-        // Compute proportional fill amounts
-        matchedFillResults = calculateMatchedFillResults(
-            leftOrder,
-            rightOrder,
-            leftOrderInfo.orderTakerAssetFilledAmount,
-            rightOrderInfo.orderTakerAssetFilledAmount
-        );
+            // Add the matchResults and the profit made during the match to the
+            // batchMatchedFillResults for this batch.
+            batchMatchedFillResults.left[matchCount] = matchResults.left;
+            batchMatchedFillResults.right[matchCount] = matchResults.right;
+            batchMatchedFillResults.profitInLeftMakerAsset = _safeAdd(
+                batchMatchedFillResults.profitInLeftMakerAsset,
+                matchResults.leftMakerAssetSpreadAmount
+            );
+            batchMatchedFillResults.profitInRightMakerAsset = _safeAdd(
+                batchMatchedFillResults.profitInRightMakerAsset,
+                matchResults.rightMakerAssetSpreadAmount
+            );
 
-        // Validate fill contexts
-        _assertValidFill(
-            leftOrder,
-            leftOrderInfo,
-            matchedFillResults.left.takerAssetFilledAmount,
-            matchedFillResults.left.takerAssetFilledAmount,
-            matchedFillResults.left.makerAssetFilledAmount
-        );
-        _assertValidFill(
-            rightOrder,
-            rightOrderInfo,
-            matchedFillResults.right.takerAssetFilledAmount,
-            matchedFillResults.right.takerAssetFilledAmount,
-            matchedFillResults.right.makerAssetFilledAmount
-        );
+            // If the leftOrder is filled, update the leftIdx, leftOrder, and leftSignature,
+            // or break out of the loop if there are no more leftOrders to match.
+            if (leftOrderInfo.orderTakerAssetFilledAmount >= leftOrder.takerAssetAmount) {
+                if (++leftIdx == leftOrders.length) {
+                    matchCount++;
+                    break;
+                } else {
+                    leftOrder = leftOrders[leftIdx];
+                    leftOrderInfo = getOrderInfo(leftOrder);
+                }
+            }
 
-        // Update exchange state
-        _updateFilledState(
-            leftOrder,
-            takerAddress,
-            leftOrderInfo.orderHash,
-            leftOrderInfo.orderTakerAssetFilledAmount,
-            matchedFillResults.left
-        );
-        _updateFilledState(
-            rightOrder,
-            takerAddress,
-            rightOrderInfo.orderHash,
-            rightOrderInfo.orderTakerAssetFilledAmount,
-            matchedFillResults.right
-        );
+            // If the rightOrder is filled, update the rightIdx, rightOrder, and rightSignature,
+            // or break out of the loop if there are no more rightOrders to match.
+            if (rightOrderInfo.orderTakerAssetFilledAmount >= rightOrder.takerAssetAmount) {
+                if (++rightIdx == rightOrders.length) {
+                    matchCount++;
```

Comment:
Won't this result in a redundant increment of `matchCount` when both orders are fully filled?

Reply:
Good point! I'll update